sdk/core/azure-core/inc/http/stream.hpp

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <algorithm>
+#include <cstdint>
+#include <cstring>
+#include <stdio.h>
+#include <vector>
+
+namespace Azure { namespace Core { namespace Http {
+
+  // BodyStream is used to read data to/from a service
+  class BodyStream {
+  public:
+    static BodyStream* null;
+
+    // Returns the length of the data; used with the HTTP Content-Length header
+    virtual uint64_t Length() = 0;
+
+    // Resets the stream back to the beginning (for retries)
+    // Derived classes that send data in an HTTP request MUST override this and implement it
+    // properly.
+    virtual void Rewind()
+    {
+      throw "Not Implemented"; // TODO: Replace with best practice as defined by guideline
+    };
+
+    // Reads more data; throws if error/canceled
+    // return copied size
+    virtual uint64_t Read(/*Context& context, */ uint8_t* buffer, uint64_t count) = 0;
+
+    // Closes the stream; typically called after all data read or if an error occurs.
+    virtual void Close() = 0;
+
+    // Desstructor. Enables derived classes to call its destructor
+    virtual ~BodyStream() = 0;
+  };
+
+  class MemoryBodyStream : public BodyStream {
+  private:
+    uint64_t m_length;
+    std::vector<uint8_t> m_buffer;
+    uint64_t offset;
+
+  public:
+    MemoryBodyStream(std::vector<uint8_t> buffer)
```

Comment:
I thought the `MemoryBodyStream` is just a wrapper of external buffer. It only takes a `char*` or `uint8_t*` as parameter, but doesn't copy the data.

Reply:
I removed the bodyBuffer from the HTTP Request.
So now the bodyBuffer needs to live someplace else right?   One option for it is that the API from service, like `uploadBlob()` have the bodyBuffer and just passes the prt, and size of it to the Stream.
The other option is that the bodyBuffer is created by some function like `CreateRequest()` and then it is moved within the stream (it used to be moved within the HTTP Request)

Reply:
okay